subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -2443,4 +2438,42 @@ public Optional<? extends RuntimeException> mapFailure(String type, Collection<T
             return DefaultConfiguration.this.mapFailure(type, failures);
         }
     }
+
+    /**
+     * Helper class that allows all configuration usage information to be stored in a byte using bitwise operations.
+     * <p>
+     * This may slightly trim the memory usage of configuration instances, which can be numerous in larger builds.
+     */
+    private enum Usage {
+        CONSUMABLE((byte) 1),
+        RESOLVABLE((byte) 2),
+        DECLARABLE_AGAINST((byte) 4),
+        DEPRECATED_FOR_CONSUMPTION((byte) 8),
+        DEPRECATED_FOR_RESOLUTION((byte) 16),
+        DEPRECATED_FOR_DECLARATION_AGAINST((byte) 32);
+
+        private static final byte EVERYTHING_ALLOWED_NOTHING_DEPRECATED = 7; // consumable & resolvable & declarable_against
```

Comment:
I feel like this would maybe be better using the actual mask values:
```suggestion
        private static final byte EVERYTHING_ALLOWED_NOTHING_DEPRECATED = (byte) (CONSUMABLE.mask | RESOLVABLE.mask | DECLARABLE_AGAINST.mask);
```
May also be worth renaming it to `EVERY_NONDEPRECATED_USAGE`?